test/ha/activator_test.go

```diff
@@ -0,0 +1,154 @@
+		t.Fatalf("Deployment %s not scaled to %d: %v", activatorDeploymentName, haReplicas, err)
+	}
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(test.ServingFlags.SystemNamespace).List(metav1.ListOptions{
+		LabelSelector: activatorLabel,
+	})
+	if err != nil {
+		t.Fatalf("Failed to get activator pods: %v", err)
+	}
+	// Sort the pods according to creation timestamp so that we can always kill the oldest one.
+	sort.Slice(pods.Items, func(i, j int) bool { return pods.Items[i].CreationTimestamp.Before(&pods.Items[j].CreationTimestamp) })
+
+	// Create first service that we will continually probe during activator restart.
+	names, resources := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey:  "1",  //make sure we don't scale to zero during the test
+			autoscaling.TargetBurstCapacityKey: "-1", // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	// Create second service that will be scaled to zero and after stopping the activator we'll
+	// ensure it can be scaled back from zero.
+	namesScaleToZero, resourcesScaleToZero := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.WindowAnnotationKey:    autoscaling.WindowMin.String(), //make sure we scale to zero quickly
+			autoscaling.TargetBurstCapacityKey: "-1",                           // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, namesScaleToZero) })
+	defer test.TearDown(clients, namesScaleToZero)
+
+	if err := e2e.WaitForScaleToZero(t, revisionresourcenames.Deployment(resourcesScaleToZero.Revision), clients); err != nil {
+		t.Fatalf("Failed to scale to zero: %v", err)
+	}
+
+	scaleToZeroURL := resources.Service.Status.URL.URL()
+	prober := test.RunRouteProber(log.Printf, clients, resources.Service.Status.URL.URL())
+	defer assertSLO(t, prober)
+
+	spoofingClient, err := pkgTest.NewSpoofingClient(clients.KubeClient, t.Logf, scaleToZeroURL.Hostname(), test.ServingFlags.ResolvableDomain, test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https))
+	if err != nil {
+		t.Fatalf("Error creating spoofing client: %v", err)
+	}
+
+	activatorPod := pods.Items[0].Name // stop the oldest activator pod
+
+	clients.KubeClient.Kube.CoreV1().Pods(test.ServingFlags.SystemNamespace).Delete(activatorPod, &metav1.DeleteOptions{
+		GracePeriodSeconds: ptr.Int64(0),
+	})
+
+	// Wait for the killed activator to disappear from the knative service's endpoints.
+	if err := waitForChangedPublicEndpoints(t, clients, resourcesScaleToZero.Revision.Name); err != nil {
+		t.Fatal("Failed to wait for the service to use only the remaining activator")
+	}
+
+	// Assert the service at the first possible moment after the killed activator disappears from its endpoints.
```

Comment:
Doesn't prober SLA kind of taking care of that? It constantly sends a stream of requests, no?

Reply:
I think it's good to check that the service works at this exact point because the enpoints have been modified. There should really be no routing problem and the service should work. I'd rather keep it.

Reply:
I realized that the original purpose was to check the service that is scaled to zero, not the one that we continually probe.